### PR TITLE
[tp] add kwargs support to prepare_module_input

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #124114

as titled, this PR adds kwargs support to PrepareModuleInput style,
where there might be modules who have only kwargs inputs but no
positional args, so we should support this

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

### CODE REVIEW

File: torch/distributed/tensor/parallel/style.py [hunk 1]
```diff
@@ -405,8 +416,32 @@ def _prepare_input_fn(self, inputs, device_mesh):
...
+                if input_layout is not None:
+                    kwarg_val = DTensor.from_local(kwarg_val, device_mesh, (input_layout,), run_check=False)
+
+            if kwarg_key in self.desired_input_kwarg_layouts:
+                assert kwarg_key in self.input_kwarg_layouts, f"input_kwarg_layouts should have the key {kwarg_key}!"
+                desired_layout = self.desired_input_kwarg_layouts[kwarg_key]
+                if desired_layout is not None and desired_layout != input_layout:
+                    kwarg_val = kwarg_val.redistribute(placements=(desired_layout,))
+
+            prepared_kwarg_inputs[kwarg_key] = kwarg_val.to_local() if self.use_local_output else kwarg_val
```
Comment (pbelevich) [inline torch/distributed/tensor/parallel/style.py:436]: `kwarg_val` can be anything(`None` or whatever type here) and it fails on HF models

Comment (wanchaol) [reply torch/distributed/tensor/parallel/style.py]: should be fixed now

File: torch/distributed/tensor/parallel/style.py [hunk 2]
```diff
@@ -372,24 +378,37 @@ class PrepareModuleInput(ParallelStyle):
     def __init__(
         self,
         *,
-        input_layouts: Union[Placement, Tuple[Placement]],
-        desired_input_layouts: Union[Placement, Tuple[Placement]],
+        input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
+        desired_input_layouts: Optional[Union[Placement, Tuple[Optional[Placement]]]] = None,
```
Comment (XilunWu) [inline torch/distributed/tensor/parallel/style.py:382]: curious: does ParallelStyle assume 1-d mesh? It seems so.

File: torch/distributed/tensor/parallel/style.py [hunk 3]
```diff
@@ -372,24 +378,37 @@ class PrepareModuleInput(ParallelStyle):
...
+        input_kwarg_layouts: Optional[Dict[str, Placement]] = None,
+        desired_input_kwarg_layouts: Optional[Dict[str, Placement]] = None,
```
Comment (XilunWu) [inline torch/distributed/tensor/parallel/style.py:384]: Can kwarg layouts include non-Tensor arguments?

**Review by XilunWu — APPROVED** [review]
lgtm. Added 2 questions. Besides, a similar logic should apply to `local_map` as well? So far `local_map` does not process `kwargs`.

File: torch/distributed/tensor/parallel/style.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/tensor/parallel/style.py:382]: yeah the TP API currently assumes 1-d mesh

**Comment by wanchaol** [comment]
> lgtm. Added 2 questions. Besides, a similar logic should apply to `local_map` as well? So far `local_map` does not process `kwargs`.

🤔 yeah I think we can interate the `local_map` API to include kwargs, does not need to be in the initial PR but it's a good thing to think about how to solve kwargs

**Pull request closed without merging**